AD review (#114)
core-wg/oscore-groupcomm: draft-ietf-core-oscore-groupcomm.md

#### Deliberately Deleted Recipient Contexts {#ssec-loss-mutable-context-overflow}

The Security Context may contain a large and variable number of Recipient Contexts. While Group OSCORE in itself does not establish a maximum number of Recipient Contexts, there are circumstances by which implementations might choose to discard Recipient Contexts or have to do so in accordance with enforced application policies. Such circumstances include the need to reclaim memory or other resources on the node hosting the endpoint, for example because the predefined maximum number of Recipient Contexts has been reached in the Security Context (see {{ssec-sender-recipient-context}}).

When a Recipient Context is deleted, this not only results in losing information about previously received messages from the corresponding other endpoint. It also results in the inability to be aware of the Security Contexts from which information has been lost.

> Therefore, if the Recipient Context is derived again from the same Security Context, there is a risk that a replayed message is not detected. If one Recipient Context has been deleted from the current Security Context, then the Replay Window of any new Recipient Context in this Security Context MUST be initialized as invalid. An exception applies when the deleted Recipient Context was created upon receiving a message and that message was not verified successfully (see {{ssec-verify-request}}, {{ssec-verify-response}}, {{sec-pairwise-verify-req}}, and {{sec-pairwise-verify-resp}}). Messages associated with a Recipient Context that has an invalid Replay Window MUST NOT be delivered to the application.

Reviewer: ```suggestion
Therefore, if the Recipient Context is derived again from the same Security Context, there is a risk that a replayed message is not detected. If any Recipient Context from any peer has ever been deleted from the current Security Context, then the Replay Window of any new Recipient Context in this Security Context MUST be initialized as invalid. An exception applies when the deleted Recipient Context was created upon receiving a message and that message was not verified successfully (see {{ssec-verify-request}}, {{ssec-verify-response}}, {{sec-pairwise-verify-req}}, and {{sec-pairwise-verify-resp}}). Messages associated with a Recipient Context that has an invalid Replay Window MUST NOT be delivered to the application.
```

Author: Suggestion taken with a minor change as below, to be consistent with phrasing used in other parts of the document. There will a commit for that.

> If any Recipient Context **associated with** any peer has ever been deleted from the current Security Context ...

Author: Fixed in https://github.com/core-wg/oscore-groupcomm/pull/114/commits/38235f76badb89d78c9f68defa534b1106e52b80

### Exhaustion of Sender Sequence Number Space {#ssec-wrap-around-partial-iv}

Since an endpoint increments its Sender Sequence Number for each new outgoing message including a Partial IV, an endpoint can eventually exhaust the Sender Sequence Number space.

Implementations MUST be able to detect an exhaustion of Sender Sequence Number space, after the endpoint has consumed the largest usable value. This may be influenced by additional limitations besides the mere 40-bit size limit of the Partial IV. If an implementation's integers support wrapping addition, the implementation MUST treat the Sender Sequence Number space as exhausted when a wrap-around is detected.

Upon exhausting the Sender Sequence Number space, the endpoint MUST NOT use this Security Context to protect further messages including a Partial IV.

> Upon (approaching) the exhaustion of the Sender Sequence Number space, the endpoint SHOULD inform the Group Manager, retrieve new Security Context parameters from the Group Manager (see {{sec-group-re-join}}), and use them to derive a new Sender Context (see {{ssec-sender-recipient-context}}). It is RECOMMENDED that the endpoint takes this course of action with some margin, i.e., well before exhausting the Sender Sequence Number space, in order to avoid a period of inability to protect messages including a Partial IV.

Reviewer: ```suggestion
When approaching the exhaustion of the Sender Sequence Number space, the endpoint SHOULD inform the Group Manager, retrieve new Security Context parameters from the Group Manager (see {{sec-group-re-join}}), and use them to derive a new Sender Context (see {{ssec-sender-recipient-context}}). It is RECOMMENDED that the endpoint takes this course of action with some margin, i.e., well before exhausting the Sender Sequence Number space, in order to avoid a period of inability to protect messages including a Partial IV.
```

The Group Manager may assign a new Sender ID to an endpoint, while leaving the Gid, Master Secret, and Master Salt unchanged in the group. In this case, the Group Manager assigns a Sender ID that has not been used in the group since the latest time when the current Gid value was assigned to the group (see {{sec-group-key-management}}).

Having retrieved the new Sender ID, and potentially other missing data for the long-term part of the Security Context, the endpoint can derive a new Sender Context (see {{ssec-sender-recipient-context}}). When doing so, the endpoint resets the Sender Sequence Number in its Sender Context to 0, and derives a new Sender Key. This is in turn used to possibly derive new Pairwise Sender Keys.

From then on, the endpoint MUST use its latest installed Sender Context to protect outgoing messages.

> The assignment of a new Sender ID may be the result of different processes. The endpoint may request a new Sender ID, e.g., because of (approaching) the exhaustion of the Sender Sequence Number space (see {{ssec-wrap-around-partial-iv}}). An endpoint may request to re-join the group, e.g., because of losing the varying part of its Security Context (see {{ssec-loss-mutable-context}}), and is provided with a new Sender ID together with the latest data for the long-term part of the Security Context.

Reviewer: ```suggestion
The assignment of a new Sender ID may be the result of different processes. The endpoint may request a new Sender ID, e.g., because of the impending exhaustion of the Sender Sequence Number space (see {{ssec-wrap-around-partial-iv}}). An endpoint may request to re-join the group, e.g., because of losing the varying part of its Security Context (see {{ssec-loss-mutable-context}}), and is provided with a new Sender ID together with the latest data for the long-term part of the Security Context.
```

   -  The client MUST store the value of the 'kid context' parameter from the request, and retain it until the long exchange is terminated. Upon establishing a new Security Context with a new Gid as ID Context (see {{new-sec-context}}), the client MUST NOT update the stored 'kid context' parameter value associated with the long exchange and the corresponding request.

>    - The client MUST store an invariant identifier of the group, which is immutable even if the Security Context of the group is re-established.  For example, this invariant identifier can be the "group name" in {{I-D.ietf-ace-key-groupcomm-oscore}}, where it is used for joining the group and retrieving the current group keying material from the Group Manager.

Reviewer: Just checking that this is intended to still say "immutable," as opposed to the term immutable which was changed to long-term?

Author: Yes, this is as intended and it is related to a different piece of information, which is actually immutable. It is not related to "parts of the Security Context" that used to be denoted as "immutable" and now are denoted as "long-term".

   - If the signature verification fails, the server SHALL stop processing the request, SHALL NOT update the Replay Window, and MAY respond with a 4.00 (Bad Request) response. Such a response MAY include an Outer Max-Age Option with value zero, and its diagnostic payload MAY contain a string, which, if present, MUST be "Decryption failed" as if the decryption of the COSE object had failed.

   - When decrypting the COSE object using the Recipient Key, the Group Encryption Algorithm from the Common Context MUST be used.

* Additionally, if the used Recipient Context was created upon receiving this request and the message is not verified successfully, the server MAY delete that Recipient Context. When this behavior is specified by the application, it mitigates attacks that aim at overloading the server's storage.

  If the server deletes the used Recipient Context in this particular circumstance, then this deletion does not require the server to initialize as invalid the Replay Window of any new Recipient Context created later within the Security Context (see {{ssec-loss-mutable-context-overflow}}).

> A server SHOULD NOT process a request if the received Recipient ID ('kid') is equal to its own Sender ID in its own Sender Context. As an exception where the requirement above is not fulfilled, the server could have deliberately prepared such a request to be sent to itself (e.g., see {{I-D.ietf-core-observe-multicast-notifications}}).

Reviewer: ```suggestion
A server SHOULD NOT process a request if the received Recipient ID ('kid') is equal to its own Sender ID in its own Sender Context. However, in some applications the server can prepare a request to be sent to itself (e.g., see {{I-D.ietf-core-observe-multicast-notifications}}), in which case such requests would be expected.
```

The Echo Option value SHOULD NOT be reused; if it is reused, it MUST be highly unlikely to have been recently used with this client. Since this response is protected with the Security Context used in the group, the client will consider the response valid upon successfully decrypting and verifying it.

The server stores the Echo Option value included in the response together with the pair (gid,kid), where 'gid' is the Group Identifier of the OSCORE group and 'kid' is the Sender ID of the client in the group. These are specified in the 'kid context' and 'kid' fields of the OSCORE Option of the request, respectively. After a group rekeying has been completed and a new Security Context has been established in the group, which results also in a new Group Identifier (see {{sec-group-key-management}}), the server MUST delete all the stored Echo values associated with members of the group.

> After receiving a 4.01 (Unauthorized) response that includes an Echo Option and originates from a verified group member, a subsequently sent client request echoing the Echo Option value MUST be sent as a unicast message to the same server.

Reviewer: ```suggestion
After receiving a 4.01 (Unauthorized) response that includes an Echo Option and originates from a verified group member, the next client request MUST echo the Echo Option value and MUST be sent as a unicast message to that server.
```

Author: After taking the suggestion into account, we thought about a different update that makes the new text clearer while still reflecting the intended behavior specified in Sections 2.3 and 2.4 of RFC 9175 (more details will be in a follow-up mail).

The updated new text reads as below. There will a commit for that.

> After receiving a 4.01 (Unauthorized) response that includes an Echo Option and originates from a verified group member, **a subsequent client request sent to that server and echoing the Echo Option value MUST be sent as a unicast message to that server.**

Author: Fixed in https://github.com/core-wg/oscore-groupcomm/pull/114/commits/3f0f2e71d8d848a126a5d79fe2944471f3e62842

* For many constrained IoT devices, it is problematic to support more than one signature algorithm. The following applies with respect to the Signature Algorithm (see {{ssec-common-context-cs-alg}}).

   Less constrained endpoints MUST implement at least one of the following and SHOULD implement both: the EdDSA signature algorithm together with the elliptic curve Ed25519 {{RFC8032}}; the ECDSA signature algorithm together with the elliptic curve P-256.

>    Constrained endpoints MUST implement at least one of the following: the EdDSA signature algorithm together with the elliptic curve Ed25519 {{RFC8032}}; the ECDSA signature algorithm together with the elliptic curve P-256.

Reviewer: Is there a definition of "less-constrained" to determine which requirement applies to a given endpoint? If not, I'd thought this would be a MUST-one/SHOULD-both (as in the previous line) and the expectation that less-constrained nodes can afford to follow the SHOULD.

Author: The following updated text should address this point while still trying to convey that expectations are higher for less constrained endpoints:

> Constrained endpoints MUST implement at least one of the following and, if affordable, SHOULD implement both: ...
>
> ...
>
> Constrained endpoints MUST implement at least one of the following ECDH curves and, if affordable, SHOULD implement both: ...

Reviewer 2: > Is there a definition of "less-constrained" to determine which requirement applies to a given endpoint? 

To answer this specific question:
We generally use RFC 7228 for a definition of "constrained node" (Section 2.1).
(This document has a -[bis][] in IOTOPS, btw.)

[bis]: https://www.ietf.org/archive/id/draft-ietf-iotops-7228bis-01.html

With respect to unprotected message fields, the following holds. First, the 'kid context' of request messages is part of the Additional Authenticated Data, making it possible to keep long exchanges active safely beyond a possible change of ID Context (Gid) following a group rekeying (see {{sec-cose-object-ext-aad}}). Second, the countersignature included in a Group OSCORE message protected in group mode is also computed over the value of the OSCORE Option, which is also part of the Additional Authenticated Data used in the signing process. This is further discussed in {{ssec-cross-group-injection}} of this document.

As discussed in {{Section 6.2.3 of I-D.ietf-core-groupcomm-bis}}, Group OSCORE addresses security attacks against CoAP listed in Sections {{11.2<RFC7252}}–{{11.6<RFC7252}} of {{RFC7252}}, especially when run over IP multicast.

The rest of this section first discusses security aspects to be taken into account when using Group OSCORE. Then it goes through aspects covered in the security considerations of OSCORE (see {{Section 12 of RFC8613}}), and discusses how they hold when Group OSCORE is used.

## Security of the Group Mode {#ssec-group-mode-security}

> The group mode defined in {{mess-processing}} relies on shared group keying material to protect communication within a group. Using the group mode has the implications discussed below. The text in this section refers to group members as the endpoints in the group storing the latest version of the group keying material.

Reviewer: I still feel like this last sentence is unclear; I'll follow up by e-mail and see what we can hash out.

Author: Thanks, the suggestion below from the mail is good. There will a commit for that.

> This section uses the term 'group members' to describe endpoints which possess the latest version of the group keying material.

Author: Fixed in https://github.com/core-wg/oscore-groupcomm/pull/114/commits/2f95812128c0f6657981a96fd457d045c2c67c68

The group mode fulfills the security properties above while also displaying the following benefits. First, the use of a Group Encryption Algorithm that does not provide integrity protection results in a minimal communication overhead, by limiting the message payload to the ciphertext without integrity tag together with the encrypted countersignature. Second, it is possible to deploy semi-trusted entities such as signature checkers (see {{sec-additional-entities}}), which can break property 5, but cannot break properties 1, 2, 3, and 4.

### Example of Need for Proof of Group Membership {#ssec-need-proof-group-membership}

> As a case in point, the importance to achieve proof of group membership is evident in a group where the Group Encryption Algorithm does not provide integrity protection; a group member leaves the group; and, after the group rekeying, associates with the group as external signature checker (see {{sec-processing-signature-checker}}). When doing so, it obtains from the Group Manager the new Signature Encryption Key, from which it can derive keystreams for encrypting and decrypting the countersignatures of messages protected in group mode.

Reviewer: ```suggestion
As a case in point, the importance of reliable proof of group membership is evident in a group where the Group Encryption Algorithm does not provide integrity protection. Suppose a group member leaves the group and, after the group rekeying, associates with the group as external signature checker (see {{sec-processing-signature-checker}}). When doing so, it obtains from the Group Manager the new Signature Encryption Key, from which it can derive keystreams for encrypting and decrypting the countersignatures of messages protected in group mode.
```

## Cross-group Message Injection {#ssec-cross-group-injection}

A given endpoint is allowed to and would likely use the same pair (private key, authentication credential) in multiple OSCORE groups, possibly administered by different Group Managers.

When a sender endpoint sends a message protected in pairwise mode to a recipient endpoint in an OSCORE group, a malicious group member may attempt to inject the message to a different OSCORE group also including the same endpoints (see {{ssec-cross-group-injection-attack}}).

> This relies on altering the content of the OSCORE Option while having the same MAC in the ciphertext that is still going to be correctly validated, which has a success probability depending on the size of the MAC.

Reviewer: ```suggestion
This relies on altering the content of the OSCORE Option in such a way that the MAC in the ciphertext will still be correctly validated, which has a success probability depending on the size of the MAC.
```

Before delivering to the application an incoming request protected in group mode that was sent over unicast, a server should carefully consider the impact that processing the request would have, without dismissing the possibility that the request reception was in fact the result of the attack discussed above. This is especially important for endpoints implementing only a silent server, since they do not support the pairwise mode and thus are able to process only requests protected in group mode. Application policies can also define specific exceptional cases where it is safe for a server to deliver such requests to the application, which can then take a final decision about whether acting on the request or not. For instance, such decision can leverage an application-level unique identifier of the server specified in the payload of the request.

> In a group where the AEAD Algorithm and Pairwise Key Agreement Algorithm are set in the Security Context, an endpoint supporting the pairwise mode SHOULD use it to protect requests sent to a single group member over unicast. As an exception where the requirement above is not fulfilled and the group mode is used to protect such requests, an endpoint could need to make the request possible to decrypt and verify for multiple other group members. An example of this exception is the method defined in {{I-D.ietf-core-observe-multicast-notifications}}.

Reviewer: ```suggestion
In a group where the AEAD Algorithm and Pairwise Key Agreement Algorithm are set in the Security Context, an endpoint supporting the pairwise mode SHOULD use it to protect requests sent to a single group member over unicast. Using the group mode to protect such requests would make the request possible to decrypt and verify for any other group member; this may sometimes be a desired property (e.g. see {{I-D.ietf-core-observe-multicast-notifications}}).
```
